src/OpenLoco/src/Map/TileManager.cpp

```diff
@@ -480,37 +490,515 @@ namespace OpenLoco::World::TileManager
+                FormatArguments::common(industry->name, industry->town);
+
+                GameCommands::setErrorText(StringIds::string_id_in_the_way);
+                break;
+            }
+        }
+    }
+
+    // 0x00462BB3
+    // If return true input el not modified
+    // If return false input el set to 0xFFFFFFFF if error text set or problem el if not set
+    static ClearResult companyAboutToBuildCheck(const World::TileElement& el)
+    {
+        auto* elSurface = el.as<SurfaceElement>();
+
+        // 0x00462C02
+        auto returnFunc = [](const CompanyId owner) -> ClearResult {
+            const auto updatingId = GameCommands::getUpdatingCompanyId();
+            if (updatingId == CompanyId::neutral || !CompanyManager::isPlayerCompany(updatingId))
+            {
+                GameCommands::setErrorText(StringIds::another_company_is_about_to_build_here);
+                return ClearResult::collisionWithErrorMessage;
+            }
+            if (owner == updatingId || CompanyManager::isPlayerCompany(owner))
+            {
+                return ClearResult::collsionNoMessage;
+            }
+            auto* company = CompanyManager::get(owner);
+            if ((company->challengeFlags & CompanyFlags::unk2) != CompanyFlags::none)
+            {
+                GameCommands::setErrorText(StringIds::another_company_is_about_to_build_here);
+                return ClearResult::collisionWithErrorMessage;
+            }
+            // Modification from vanilla
+            company->challengeFlags |= CompanyFlags::unk1;
+            return ClearResult::noCollision;
+        };
+
+        if (elSurface == nullptr)
+        {
+            CompanyId owner = getTileOwner(el);
+            if (owner == CompanyId::null)
+            {
+                return ClearResult::noCollision;
+            }
+
+            return returnFunc(owner);
+        }
+        else
+        {
+            auto* elp = &el;
+            while (!elp->isLast())
+            {
+                elp = elp->next();
+                if (!elp->isFlag5())
+                {
+                    continue;
+                }
+                const auto owner = getTileOwner(*elp);
+                if (owner == CompanyId::null)
+                {
+                    continue;
+                }
+
+                return returnFunc(owner);
+            }
+            GameCommands::setErrorText(StringIds::another_company_is_about_to_build_here);
+            return ClearResult::collisionWithErrorMessage;
+        }
+    }
+
+    enum class Sub462B4FResult
+    {
+        allCollisionsRemoved,
+        collision,
+        collisionErrorSet,
+        noCollision,
+        collisionRemoved,
+    };
+
+    // 0x00462937
+    static bool canConstructAtWithClear(const World::Pos2& pos, uint8_t baseZ, uint8_t clearZ, const QuarterTile& qt, uint8_t flags, std::function<ClearResult(const TileElement& el)> clearFunc)
```

Comment:
Should we expose this already, eyeing future implementations? Presumably they wouldn't use the legacy functions any more.

Reply:
This function isn't actually ever called from outside of this file. There are 3 calling functions for it though. I was thinking we would have 5 calling functions. 2 legacy ones (there is one setup that doesn't require a legacy) and 3 new ones.